programs/utilities/block_counter.sc

```diff
@@ -1,28 +1,151 @@
+//Counts blocks in a given region, optionally filtering by block type or tag. 
+//Prints the output into a chat or a file.
+//If vertical scan is set to true, it will return a separate report for each y value.
+//By gnembon and Firigion
+
 __config() ->
 {
    'commands' -> {
+      '' -> 'help',
       '<from_pos> <to_pos>' -> 'count_blocks',
       '<from_pos> <to_pos> <filter>' -> 'count_blocks_filter',
-   }
+      '<from_pos> <to_pos> <blockpredicate>' -> 'count_blocks_predicate',
+      'save' -> ['save_scan', null],
+      'save <name>' -> 'save_scan',
+      'autosave <bool>' -> 'autosave',
+      'layered_scan <bool>' -> 'vertical_scan', // I'm keeping this as an option because doing the vertically spearated scans is 30% to 100% slower
+      'histogram <block>' -> 'histogram',
+      'tally_level <int>' -> 'tally_level',
+      //'book' -> 'get_book_report',
+      'last' -> 'last_report',
+   },
+   'allow_command_conflicts' -> true
 };
 
+
+/////////////////////////////
+///// Scans
+/////////////////////////////
+
 count_blocks(from_pos, to_pos) ->
 (
-    stats = {};
-    volume(from_pos, to_pos, stats:str(_) += 1 );
-    tally(stats, sum(... values(stats)));
+    global_data = make_data_stuct(from_pos, to_pos, null); //make global empty data structure
+    stats = global_data:'stats'; //save reference for easier access
+
+    if(global_vertscan,
+      // do a scan separating results by y value
+      area = global_data:'area';
+      [miny, maxy] = get_minmax_pos(from_pos, to_pos);
+      total = 0;
+
+      volume(from_pos, to_pos, 
+         stats:(miny+floor(total/area)):str(_) += 1; //very slightly faster than calling pos(_):1
+         total += 1; 
+      );
+      vertical_report(),
+
+      // do a regular scan
+      volume(from_pos, to_pos, stats:str(_) += 1 );
+      tally(global_data:'stats', global_data:'volume');
+      global_data:'total' = sum(... values(global_data:'stats'));
+    );
+
+    if(global_autosave, save_scan(null));
 );
 
 count_blocks_filter(from_pos, to_pos, filtr) ->
 (
-    stats = {}; total = 0;
-    volume(from_pos, to_pos, total += 1; if (_~filtr, stats:str(_) += 1 ) );
-    tally(stats, total);
+    global_data = make_data_stuct(from_pos, to_pos, filtr); //make global empty data structure
+    stats = global_data:'stats'; //save reference for easier access
+
+    if(global_vertscan,
+      // do a scan separating results by y value
+      area = global_data:'area';
+      [miny, maxy] = get_minmax_pos(from_pos, to_pos);
+      total = 0;
+
+      volume(from_pos, to_pos, 
+         if (_~filtr, stats:(miny+floor(total/area)):str(_) += 1); //very slightly faster than calling pos(_):1
+         total += 1; 
+      );
+      vertical_report(),
+      // do a regular scan
+      volume(from_pos, to_pos, total += 1; if (_~filtr, stats:str(_) += 1 ) );
+      tally(global_data:'stats', global_data:'volume');
+      global_data:'total' = sum(... values(global_data:'stats'));
```

Comment:
wait, I may have gotten lost in the code here, but why not just set this to be the `total `variable that you already have? ik that you seem to be incrementing the total each time instead of only when the block matches the filter, but u also don't seem to be doing much with that variable (unless I'm mistaken), so you could just use that no? It would be faster than unpacking and summing after you've done your calculations